components/org.wso2.carbon.identity.api.server.organization.user.invitation.management/org.wso2.carbon.identity.api.server.organization.user.invitation.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/user/invitation/management/v1/core/GuestApiServiceCore.java

```diff
@@ -0,0 +1,253 @@
+import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
+import static javax.ws.rs.core.Response.Status.NOT_IMPLEMENTED;
+
+/**
+ * Service class for the Shared User Invitation Management APIs.
+ */
+public class GuestApiServiceCore {
+
+    private static List<RoleAssignmentResponse> buildRoleAssignmentResponse(Invitation invitationRecord) {
+
+        List<RoleAssignmentResponse> roleAssignmentResponseList = new ArrayList<>();
+        for (RoleAssignments roleAssignment : invitationRecord.getRoleAssignments()) {
+            if (roleAssignment.getRoles() != null) {
+                RoleAssignmentResponse roleAssignmentResponse = new RoleAssignmentResponse();
+                roleAssignmentResponse.setApplicationId(roleAssignment.getApplicationId());
+                roleAssignmentResponse.setRoles(Arrays.asList(roleAssignment.getRoles()));
+                roleAssignmentResponseList.add(roleAssignmentResponse);
+            }
+        }
+        return roleAssignmentResponseList;
+    }
+
+    /**
+     * Creates the invitation to the shared user.
+     *
+     * @param invitationRequestBody Contains the details of the invitation.
+     * @return The details of the created invitation.
+     */
+    public InvitationSuccessResponse createInvitation(InvitationRequestBody invitationRequestBody) {
+
+        InvitationCoreServiceImpl invitationCoreService = new InvitationCoreServiceImpl();
+        Invitation invitation = new Invitation();
+        invitation.setUsername(invitationRequestBody.getUsername());
+        invitation.setUserDomain(invitationRequestBody.getUserDomain());
+        invitation.setUserRedirectUrl(invitationRequestBody.getUserRedirectUrl());
+        if (invitationRequestBody.getRoleAssignments() != null) {
+            List<RoleAssignments> roleAssignments = new ArrayList<>();
+            for (RoleAssignmentRequestBody roleAssignmentRequestBody : invitationRequestBody.getRoleAssignments()) {
+                if (roleAssignmentRequestBody.getRoles() != null) {
+                    RoleAssignments roleAssignment = new RoleAssignments();
+                    roleAssignment.setApplicationId(roleAssignmentRequestBody.getApplicationId());
+                    roleAssignment.setRoles(roleAssignmentRequestBody.getRoles().toArray(new String[0]));
+                    roleAssignments.add(roleAssignment);
+                }
+            }
+            invitation.setRoleAssignments(roleAssignments.toArray(new RoleAssignments[0]));
+        }
+        Invitation invitationResponse;
+        try {
+            invitationResponse = invitationCoreService.createInvitation(invitation);
+        } catch (UserInvitationMgtException e) {
+            if (ERROR_CODE_USER_NOT_FOUND.getCode().equals(e.getErrorCode())) {
+                throw handleException(BAD_REQUEST, SharedUserInvitationMgtConstants.ErrorMessage
+                        .ERROR_CODE_USER_NOT_FOUND, invitation.getUsername());
+            }
```

Comment:
Can't we first catch client errors and throw is and then catch server errors?

Reply:
Check other places as well

Reply:
Done